### ISSUE TITLE
support for multi value return

### ISSUE THREAD

**Issue by kwerner8** [grain-lang/grain#1456]
Is it possible to compile an exported function in Grain to wasm which returns multiple values?

I want to compile the following function to WebAssembly and then call it in Wasmtime:

```
export let multi = (x,y) => {
    (x+y, x-y, x*y)
}
```
Currently this function compiles to:

```
func (;153;) (param i32 i32) (result i32)

... WebAssembly code ...
```

So it just returns one value but it should return three values like this:
```
func (;153;) (param i32 i32) (result i32 i32 i32)

... WebAssembly code ..
```

Does anybody know how to fix this problem? I already compiled other languages like C or Rust to WebAssembly. For those I could enable an option like _--enable-multi-value_ for the compilation and I got the right result. Is there something similar for Grain?

**Comment by ospencer** [comment]
Not at the moment! Grain doesn't get have first-class multivalue support, but it's something we plan to support in the future.

**Comment by phated** [comment]
Something to note: NEAR doesn't support multivalue so we needed to make sure output **didn't** contain multivalues when some optimizations were producing them. We're going to revisit this in the future because dropping NEAR support would likely be considered a breaking change.

**Comment by kwerner8** [comment]
My goal is to write a function in Grain and then call the wasm file in Wasmtime. There I can use the function from Grain as an export of the wasm file. Therefore, a multi-value return value would be very helpful to pass multiple values from the exported function back to the Wasmtime host.

**Comment by kwerner8** [comment]
My goal is to write a function in some language and then compile it into a wasm file (it can be any language that can be compiled into a suitable wasm file). Therefore, I looked into many programming languages and I also came across Grain. I found it very interesting to look into it in more detail. 
Then I want to use this function as an export in a host code. This host code must be written in C/C++ for various reasons that cannot be changed. Therefore, I also cannot use `wit-bindgen` because a [host in C/C++](https://github.com/bytecodealliance/wit-bindgen#contributing-bindings) is not supported yet.

So far, I have called the exported function directly and passed the input to it, similar to this [example](https://docs.wasmtime.dev/examples-c-gcd.html). I used the Wasmtime runtime. Therefore, I was investigating the multi-value return because it simplifies passing data between the wasm-module and the host.

**Comment by jzabinski-dolios** [comment]
Just to throw in another use case:

I am building an astrology app. One of the app's features is the location of seven planets in the sky. To calculate the positions of those planets, it is wonderfully efficient to use WebAssembly on the client side. The positions of the seven _as a set_ is always interesting; the position of any one planet is never interesting. Instead of asking for all seven planets separately, it is more performant to ask once, and get back seven answers.

The input for the WebAssembly function is a time, represented as an f64. The output is ideally a set of seven f64's. Currently, though, because multi-value is not implemented, the app has to know a lot about data representation so that it can receive a pointer and retrieve the data in a relevant way.

**Comment by ospencer** [comment]
Yes, there's current work in progress on adding support for Grain.

**Comment by ospencer** [comment]
I've started a discussion at https://github.com/grain-lang/grain/discussions/1650 for tracking, and we'll discuss all updates there 🙂
